### PR TITLE
Delete registration for virtual guest deleter

### PR DESCRIPTION
**Pull request by zhanggbj**
Hi @jianqiu ,

Please take a look if you get time. I'll then summit to master after PR #184 is merged. Thanks

    1.Delete registration
    2.Add related unit test

[#138226191](https://www.pivotaltracker.com/story/show/138226191)

/cc @mattcui

### CODE REVIEW

File: softlayer/common/fakes/fake_vm.go [hunk 1]
```diff
@@ -141,6 +141,12 @@ type FakeVM struct {
 	updateAgentEnvReturns struct {
 		result1 error
 	}
+	DeleteAgentEnvStub        func() error
+	deleteAgentEnvMutex       sync.RWMutex
+	deleteAgentEnvArgsForCall []struct{}
+	deleteAgentEnvReturns     struct {
+		result1 error
```
Comment (maximilien) [inline softlayer/common/fakes/fake_vm.go]: What's this `result1` business? Can we use a better name? Is this auto-generated?

File: softlayer/vm/softlayer_virtual_guest_deleter.go [hunk 2]
```diff
@@ -40,6 +42,16 @@ func (c *softLayerVMDeleter) Delete(cid int) error {
 		}
 	}
 
+	vm, found, err := c.vmFinder.Find(cid)
+	if err != nil || !found {
+		return bosherr.WrapErrorf(err, "Cannot find VirtualGuest with id: %d.", cid)
```
Comment (maximilien) [inline softlayer/vm/softlayer_virtual_guest_deleter.go]: Can we separate these two cases? I.e.,

1. when there is `err`
2. when `!found`

I think that will make code easier to read and most importantly to test.

File: softlayer/vm/softlayer_virtual_guest_deleter_test.go [hunk 3]
```diff
@@ -2,27 +2,33 @@ package vm_test
 
 import (
 	. "github.com/cloudfoundry/bosh-softlayer-cpi/softlayer/common"
+	fakescommon "github.com/cloudfoundry/bosh-softlayer-cpi/softlayer/common/fakes"
 	slh "github.com/cloudfoundry/bosh-softlayer-cpi/softlayer/common/helper"
```
Comment (maximilien) [inline softlayer/vm/softlayer_virtual_guest_deleter_test.go]: please arrange these imports, something like:

```
"time"

slh "github.com/cloudfoundry/bosh-softlayer-cpi/softlayer/common/helper"
boshlog "github.com/cloudfoundry/bosh-utils/logger"
...

fakescommon "github.com/cloudfoundry/bosh-softlayer-cpi/softlayer/common/fakes"
...
. "github.com/onsi/ginkgo"
...

```

File: softlayer/vm/softlayer_virtual_guest_deleter_test.go [hunk 4]
```diff
@@ -58,13 +70,45 @@ var _ = Describe("SoftlayerVirtualGuestDeleter", func() {
 			})
 		})
 
+		Context("when can not find virtual guest with id 1234567", func() {
```
Comment (maximilien) [inline softlayer/vm/softlayer_virtual_guest_deleter_test.go]: I'd expect to see a fail tests for when not found and when err...

File: softlayer/vm/softlayer_virtual_guest_deleter_test.go [hunk 5]
```diff
@@ -58,13 +70,45 @@ var _ = Describe("SoftlayerVirtualGuestDeleter", func() {
...
+			})
+
+			It("returns an error", func() {
+				Expect(err).To(HaveOccurred())
+				Expect(fakeVm.DeleteAgentEnvCallCount()).To(Equal(1))
+				Expect(err.Error()).To(ContainSubstring("Deleting VM's agent env"))
+			})
+		})
+
 		Context("when deleting object and error occures", func() {
```
Comment (maximilien) [inline softlayer/vm/softlayer_virtual_guest_deleter_test.go]: typo "... occurs"

File: softlayer/common/fakes/fake_vm.go (hunk 1 above)
Comment (zhanggbj) [inline softlayer/common/fakes/fake_vm.go]: Yes, this is generated by counterfeiter.

File: softlayer/vm/softlayer_virtual_guest_deleter.go (hunk 2 above)
Comment (zhanggbj) [inline softlayer/vm/softlayer_virtual_guest_deleter.go]: Done.

File: softlayer/vm/softlayer_virtual_guest_deleter_test.go (hunk 3 above)
Comment (zhanggbj) [inline softlayer/vm/softlayer_virtual_guest_deleter_test.go]: Done.

File: softlayer/vm/softlayer_virtual_guest_deleter_test.go (hunk 4 above)
Comment (zhanggbj) [inline softlayer/vm/softlayer_virtual_guest_deleter_test.go]: Done.

File: softlayer/vm/softlayer_virtual_guest_deleter_test.go (hunk 5 above)
Comment (zhanggbj) [inline softlayer/vm/softlayer_virtual_guest_deleter_test.go]: Done.

File: softlayer/common/fakes/fake_vm.go (hunk 1 above)
Comment (maximilien) [inline softlayer/common/fakes/fake_vm.go]: Cool, thx

**Comment by maximilien** [comment]
I am done with my review. 

@mattcui or @jianqiu please double check and note here and I'll merge.

**Comment by zhanggbj** [comment]
Hi Dr.Max @maximilien, 

This commit is just for PR review as dependency (PR #184) is still in review. May we can merge it to develop_v3 firstly and cherry-pick to master later. Just a reminder. Thanks.

**Comment by maximilien** [comment]
OK @zhanggbj still waiting on @jianqiu to fix rebase issues and address comments. Best.

**Comment by maximilien** [comment]
@zhanggbj What is the status of this PR? Shall I review again?

**Comment by maximilien** [comment]
@zhanggbj still waiting on this?

**Comment by zhanggbj** [comment]
Yes, still pending on PR #184 .

**Comment by zhanggbj** [comment]
Dependent PR and this one are all merged. So close this PR. Thanks.

**Pull request closed without merging**